### ISSUE TITLE
alternating results running grunt qunit

### ISSUE THREAD

**Issue by mfrasca** [gruntjs/grunt-contrib-qunit#139]
after so many years working with Pascal, C/C++, Python, with unit testing and test coverage monitoring, I feel like a beginner again! I'm very new to all the tools I'm using now, so I'm sure I'm missing some very basic points.

I'm writing tests for the front-end of [Ghini/ghini.web](https://github.com/Ghini/ghini.web/), using [this Gruntfile](https://github.com/Ghini/ghini.web/blob/master/Gruntfile.js), and I have this funny alternating failure/success pattern:

```
mario@rune:~/Local/github/Ghini/ghini.web$ grunt qunit
Running "qunit:files" (qunit) task
Testing public/test/test.html >> Running qunit.spawn: public/test/test.html
>> Running qunit.begin
>> Running qunit.moduleStart:
>> Running qunit.testStart: Object.values
>> PhantomJS error:
>>  TypeError: undefined is not an object (evaluating 'QUnit.jsDump.parse') 
>>  [ { file: 'phantomjs://code/bridge.js', line: 40, function: '' },
>>   { file: 'file:///home/mario/Local/github/Ghini/ghini.web/public/test/qunit.js',
>>     line: 882,
>>     function: 'runLoggingCallbacks' },
>>   { file: 'file:///home/mario/Local/github/Ghini/ghini.web/public/test/qunit.js',
>>     line: 1283,
>>     function: 'pushFailure' },
>>   { file: 'file:///home/mario/Local/github/Ghini/ghini.web/public/test/qunit.js',
>>     line: 1042,
>>     function: 'run' },
>>   { file: 'file:///home/mario/Local/github/Ghini/ghini.web/public/test/qunit.js',
>>     line: 1214,
>>     function: '' },
>>   { file: 'file:///home/mario/Local/github/Ghini/ghini.web/public/test/qunit.js',
>>     line: 2271,
>>     function: 'process' },
>>   { file: 'file:///home/mario/Local/github/Ghini/ghini.web/public/test/qunit.js',
>>     line: 2254,
>>     function: 'begin' },
>>   { file: 'file:///home/mario/Local/github/Ghini/ghini.web/public/test/qunit.js',
>>     line: 2215,
>>     function: '' } ]
>> PhantomJS error:
>>  TypeError: undefined is not an object (evaluating 'QUnit.jsDump.parse') 
>>  [ { file: 'phantomjs://code/bridge.js', line: 40, function: '' },
>>   { file: 'file:///home/mario/Local/github/Ghini/ghini.web/public/test/qunit.js',
>>     line: 882,
>>     function: 'runLoggingCallbacks' },
>>   { file: 'file:///home/mario/Local/github/Ghini/ghini.web/public/test/qunit.js',
>>     line: 1283,
>>     function: 'pushFailure' },
>>   { file: 'file:///home/mario/Local/github/Ghini/ghini.web/public/test/qunit.js',
>>     line: 1397,
>>     function: 'pushFailure' },
>>   { file: 'file:///home/mario/Local/github/Ghini/ghini.web/public/test/qunit.js',
>>     line: 2039,
>>     function: 'onerror' } ]

>> Running qunit.fail.timeout: undefined
>> PhantomJS timed out, possibly due to a missing QUnit start() call.
Warning: 1/1 assertions failed (0ms) Use --force to continue.

Aborted due to warnings.
```

I don't change anything, I just run it again, and I can do so immediately or after 10s, it makes no difference:
```
mario@rune:~/Local/github/Ghini/ghini.web$ grunt qunit
Running "qunit:files" (qunit) task
Testing public/test/test.html >> Running qunit.spawn: public/test/test.html
>> Running qunit.begin
>> Running qunit.moduleStart:
>> Running qunit.testStart: Object.values
.>> Running qunit.testDone: Object.values
>> Running qunit.testStart: String.formatU
.>> Running qunit.testDone: String.formatU
>> Running qunit.testStart: String.formatU-decimal_number
.>> Running qunit.testDone: String.formatU-decimal_number
>> Running qunit.testStart: parse_hash-garden
.>> Running qunit.testDone: parse_hash-garden
>> Running qunit.testStart: parse_hash-map
.>> Running qunit.testDone: parse_hash-map
>> Running qunit.testStart: parse_hash-garden-map
.>> Running qunit.testDone: parse_hash-garden-map
>> Running qunit.testStart: parse_hash-ignore
.>> Running qunit.testDone: parse_hash-ignore
>> Running qunit.testStart: shorten
.>> Running qunit.testDone: shorten
>> Running qunit.testStart: set_alternative
.>> Running qunit.testDone: set_alternative
>> Running qunit.moduleDone:
OK
>> 37 assertions passed (81ms)
>> Coverage:
>> -  Lines: 21.59%
>> -  Statements: 21.33%
>> -  Functions: 12.66%
>> -  Branches: 10.91%

Done.
```

the only differences with the checked in Gruntfile are the following:

```
diff --git a/Gruntfile.js b/Gruntfile.js
index 7c7ccc4..b38957b 100644
--- a/Gruntfile.js
+++ b/Gruntfile.js
@@ -40,4 +40,28 @@ module.exports = function(grunt) {
     
     // Task to run tests
     grunt.registerTask('test', 'qunit');
+    grunt.event.on('qunit.spawn', function (url) {
+        grunt.log.ok('Running qunit.spawn: ' + url);
+    });
+    grunt.event.on('qunit.fail.load', function (url) {
+        grunt.log.ok('Running qunit.fail.load: ' + url);
+    });
+    grunt.event.on('qunit.begin', function () {
+        grunt.log.ok('Running qunit.begin');
+    });
+    grunt.event.on('qunit.moduleStart', function (name) {
+        grunt.log.ok('Running qunit.moduleStart: ' + name);
+    });
+    grunt.event.on('qunit.testDone', function (name) {
+        grunt.log.ok('Running qunit.testDone: ' + name);
+    });
+    grunt.event.on('qunit.testStart', function (name) {
+        grunt.log.ok('Running qunit.testStart: ' + name);
+    });
+    grunt.event.on('qunit.moduleDone', function (name) {
+        grunt.log.ok('Running qunit.moduleDone: ' + name);
+    });
+    grunt.event.on('qunit.fail.timeout', function (url) {
+        grunt.log.ok('Running qunit.fail.timeout: ' + url);
+    });
 };
```

**Comment by Arkni** [comment]
`grunt-qunit-istanbul` normally works as a drop-in replacement for `grunt-contrib-qunit` task, so no need to load [`grunt-contrib-qunit`](https://github.com/Ghini/ghini.web/blob/f91e4a7bc6b94db67e0797ec190eb40166e9872c/Gruntfile.js#L38) in your code. Plus, the config of your task are not part of this package.

Note that `grunt-qunit-istanbul` is not up-to-date and will not work with `QUnit > 2.x`, because `QUnit.jsDump` was removed as of `2.0.0` (See [`QUnit 2.x Upgrade Guide`](http://qunitjs.com/upgrade-guide-2.x/#replace-qunit-jsdump-with-qunit-dump)).

Another approach you can follow (in case you need test coverage) is this one used in `QUnit` itself: https://github.com/qunitjs/qunit/pull/1079

**Comment by mfrasca** [comment]
what do you think? much work to update `grunt-qunit-instanbul` to QUnit 2? I tried to do the same as in `grunt-contrib-qunit/phantomjs/bridge.js` (grab `dump` or `jsDump` from QUnit), but now I alternate between success and a different failure. and yes, I only need test coverage (publishing), so I might try `QUnit` alone, if it can do that.

```
mario@rune:~/Local/github/Ghini/ghini.web$ grunt qunit
Running "qunit:files" (qunit) task
Testing public/test/test.html FFFFFFFFF
>> Object.values
>> Message: Died on test #1 global code@file:///home/mario/Local/github/Ghini/ghini.web/public/test/MyJsUnitTests.js:1:11: undefined is not a function (evaluating 'Object.values(l)')
>> Actual: null
>> Expected: undefined
>> file:///home/mario/Local/github/Ghini/ghini.web/public/test/MyJsUnitTests.js:4:35

>> String.formatU
>> Message: Died on test #1 global code@file:///home/mario/Local/github/Ghini/ghini.web/public/test/MyJsUnitTests.js:9:11: undefined is not a function (evaluating '"abc.{0}-{1}".formatU([5, 6])')
>> Actual: null
>> Expected: undefined
>> file:///home/mario/Local/github/Ghini/ghini.web/public/test/MyJsUnitTests.js:10:39

>> String.formatU-decimal_number
>> Message: Died on test #1 global code@file:///home/mario/Local/github/Ghini/ghini.web/public/test/MyJsUnitTests.js:14:11: undefined is not a function (evaluating '"abc.{0}".formatU([5.1])')
>> Actual: null
>> Expected: undefined
>> file:///home/mario/Local/github/Ghini/ghini.web/public/test/MyJsUnitTests.js:15:35

>> parse_hash-garden
>> Message: Died on test #1 global code@file:///home/mario/Local/github/Ghini/ghini.web/public/test/MyJsUnitTests.js:26:11: Can't find variable: parse_hash
>> Actual: null
>> Expected: undefined
>> file:///home/mario/Local/github/Ghini/ghini.web/public/test/MyJsUnitTests.js:27:32

>> parse_hash-map
>> Message: Died on test #1 global code@file:///home/mario/Local/github/Ghini/ghini.web/public/test/MyJsUnitTests.js:31:11: Can't find variable: parse_hash
>> Actual: null
>> Expected: undefined
>> file:///home/mario/Local/github/Ghini/ghini.web/public/test/MyJsUnitTests.js:32:32

[et cetera]
```

**Comment by Arkni** [comment]
Do you use any polyfill to shim `Object.values()`, as this method is part of `ES2017` and `PhantomJS` doesn't support that (the current stable version `2.1.1`) ?

**Comment by mfrasca** [comment]
What is a “polyfill” ? I have a definition for `Object.values()` in  my "cuchubo.js" file, is that it? Anyhow, this is the reason I am testing the function. once it works, the next not, once ok, next not, and so on.

```
Object.values = function(obj) {
    // Object.values({key: 'value'}) => ['value']
    return Object.keys(obj).map(function(key) {return obj[key];});
};
```

**Comment by mfrasca** [comment]
I don't think that is the point: I can make it optional as with 
```
if (Object.values === undefined) 
    Object.values = function(obj) {
        // Object.values({key: 'value'}) => ['value']
        return Object.keys(obj).map(function(key) {return obj[key];});
    };
```

and I still have the same alternating success failure behaviour, with `Object.value` as well as with all other functions I'm testing.

**Comment by mfrasca** [comment]
you know what is interesting now. I removed the “polyfill” and I got one failing test: I ran `grunt qunit` and I got that single failure, plus all other successes. No alternating behaviour any more. Tried a couple of times and no variation: that single one failing test. Now I fix the failing test, run `grunt qunit` again and indeed all tests succeed. Next time: that funny failure on **all** tests.

**Comment by Arkni** [comment]
I just cloned your repo and commented out the line:
```js
grunt.loadNpmTasks('grunt-qunit-istanbul');
```
and then I run `grunt qunit` and everything works as expected. Now, you have to take a look at the approach I linked in my comment https://github.com/gruntjs/grunt-contrib-qunit/issues/139#issuecomment-279217859 to integrate coverage.

**Comment by mfrasca** [comment]
woops, shame on me and my eyes! I had reacted to your initial comment "remove the useless grunt.loadNpmTasts" by removing the other one. :roll_eyes:

**Comment by mfrasca** [comment]
woops... this does work, but wait, you are suggesting I keep `grunt-contrib-qunit`, not `grunt-qunit-istanbul`?
I see, I was attempting the opposite way. I'll do my homework and come back if I _really_ don't manage. thanks for your time.

**Comment by Arkni** [comment]
Just ping me and will try to help as much as I can.

**Comment by mfrasca** [comment]
well, I have the impression that I have to copy a HUGE amount of configuration stuff from [that Gruntfile](https://raw.githubusercontent.com/trentmwillis/qunit/346a0d9492ae5e8bc719e51492241c7f06b69944/Gruntfile.js) into mine, something I did not need with `grunt-qunit-istanbul`. and it's configuration stuff I don't follow, so I would rather fix `grunt-qunit-istanbul`... I'll try a bit harder, and come back later.

**Comment by mfrasca** [comment]
I found a rather simple solution, still using `grunt-qunit-istanbul` and my Gruntfile.js.
simply `mkdir temp` before running `grunt qunit`. the directory is removed after successfully running the tests.

**Comment by mfrasca** [comment]
this solves it for me, no need to change anything else.
```
--- /home/mario/Local/github/Ghini/ghini.web/node_modules/grunt-qunit-istanbul/tasks/qunit.js	2017-02-12 20:09:05.122033295 -0500
+++ /tmp/qunit.js	2017-02-12 20:12:07.122035314 -0500
@@ -237,12 +237,10 @@
         // clear instrumented files folder & generate a new one
         if (options.coverage && options.coverage.instrumentedFiles) {
           if (!fs.existsSync(options.coverage.instrumentedFiles)) {
-            grunt.file.mkdir(options.coverage.instrumentedFiles);
-            options.transport.instrumentedFiles = options.coverage.instrumentedFiles;
-          } else {
-            options.transport.instrumentedFiles = fs.realpathSync(options.coverage.instrumentedFiles);
-            rimraf.sync(options.transport.instrumentedFiles);
+            fs.mkdirSync(options.coverage.instrumentedFiles);
           }
+          options.transport.instrumentedFiles = fs.realpathSync(options.coverage.instrumentedFiles);
+          rimraf.sync(options.transport.instrumentedFiles);
         }
 
         // write instrumented file information to an temporary file
```

**Comment by mfrasca** [comment]
Sorry, I really had not understood the relation between gruntjs/grunt-contrib-qunit and asciidisco/grunt-qunit-istanbul, and now that I think I got it, I would so much like that asciidisco/grunt-qunit-istanbul be merged into gruntjs/grunt-contrib-qunit.
Any chances?

**Comment by Arkni** [comment]
Doing so will diverge from the goal of this plugin which is _running `QUnit` tests in `PhantomJS`_.
`asciidisco/grunt-qunit-istanbul` was an attempt (and it succeed) to integrate `Istanbul` with this plugin, but right now, that plugin is not maintained anymore.

Merging the two plugin will cause a lot of headache. We first need to merge `grunt-lib-phantomjs` and `grunt-lib-phantomjs-istanbul`, and then merging the two other ones.
This is not an easy option and there are no plans, as far as I know, to do the merge or to integrate test coverage.

tbh, I still see the solution used by `QUnit` core team as a proper one, this will save you a lot of time instead of just trying to sync the fork with upstream.

**Comment by mfrasca** [comment]
> I still see the solution used by QUnit core team as a proper one

I get your point and I am already seeing the headache you refer to, so I hope you do agree we miss basic documentation. :books: 
I come to JavaScript from other environments, I am used to reading docs, following them, understanding what I'm doing, and see things work quickly. The task at hand is so basic, isn't it? "configure test coverage and publish it on coveralls", and I'm lost in too many details of an environment I don't master. :person_fencing: 
My ideal situation would be "just use one (extra) thing" and "these is the 5 lines configuration you need". :neutral_face: 
Maybe other unit testing suites make it easier?

**Comment by Arkni** [comment]
I know documentation is an important thing, but for plugins like this, we only need to document the options and nothing more.

But, I do understand your struggle. I would like to know which part we need to document and I will try to document it as soon as possible.

---

For coverage, there are plugin that do that: [`grunt-istanbul`](https://github.com/taichi/grunt-istanbul) for creating and reporting coverage and [`grunt-coveralls`](https://github.com/pimterry/grunt-coveralls) for loading coverage results and submit them to `Coveralls.io`.

And in case you want to use another unit testing framework, you can take a look at [`Jasmine`](https://jasmine.github.io/)  ([`Grunt` plugin](https://github.com/gruntjs/grunt-contrib-jasmine) and [`coverage` template](https://github.com/maenu/grunt-template-jasmine-istanbul))

**Comment by Arkni** [comment]
I'll have to close this issue. Feel free to continue the discussion, though :)

**Comment by mfrasca** [comment]
I have a nodejs application, so I have both a server and a client program, but let's simplify my use case to just testing the client code. To me, it sounds like a very good example on how to configure this tool.

- very small program, registered on github as `<user>/<repo>.git`
- all sources are in `public/js/*.js`
- bit by bit gathering tests in `test/test_*.js`
- publishing test results on `travis-ci.org/<user>/<repo>`
- publishing coverage reports on `coveralls.io/github/<user>/<repo>`

My focus would be on having the report published on coveralls, and that this be done by travis.

**Comment by mfrasca** [comment]
oh, I'm fine with editing the docs, too, once I know what I'm missing there and what to write.

**Comment by mfrasca** [comment]
so as of now I have my unit tests nicely running , but no coverage information being collected.
```
mario@rune:~/Local/github/Ghini/ghini.web$ grunt test
Running "qunit:files" (qunit) task
Testing public/test/test.html .........OK
>> 9 tests completed with 0 failed, 0 skipped, and 0 todo. 
>> 37 assertions (in 70ms), passed: 37, failed: 0

Done.
```
you said I should install `grunt-istanbul` and `grunt-coveralls`, I will see how that goes.

**Comment by mfrasca** [comment]
now I have installed them, and read and followed (but likely not completely understood) https://www.npmjs.com/package/grunt-istanbul, and my Gruntfile.js is not much more than a copy of the example.

```js
module.exports = function(grunt) {
    // Project configuration.
    grunt.initConfig({
        qunit: {
            files: ['test/test.html'],
            options: {
                reporter: 'spec',
                '--web-security': 'no'
            }
        },
        env: {
            coverage: {
                APP_DIR_FOR_CODE_COVERAGE: '../test/coverage/instrument/public/'
            }
        },
        instrument: {
            files: 'public/js/*.js',
            options: {
                lazy: true,
                basePath: 'test/coverage/instrument/'
            }
        },
        storeCoverage: {
            options: {
                dir: 'test/coverage/reports'
            }
        },
        makeReport: {
            src: 'test/coverage/reports/**/*.json',
            options: {
                type: 'lcov',
                dir: 'test/coverage/reports',
                print: 'detail'
            }
        },
        coveralls: {
            options: { // Options relevant to all targets 
                force: false
            },
            qunit: { // 
                src: 'report/lcov/lcov.info', // LCOV coverage file (can be string, glob or array) 
                options: { // Any options for just this target 
                }
            }
        }
    });

    // Load plugins
    grunt.loadNpmTasks('grunt-contrib-qunit');
    grunt.loadNpmTasks('grunt-env');
    grunt.loadNpmTasks('grunt-istanbul');
    grunt.loadNpmTasks('grunt-coveralls');
    
    // Task to run tests
    grunt.registerTask('test', 'qunit');
    grunt.registerTask('coverage', ['env:coverage', 'instrument', 'test',
                                    'storeCoverage', 'makeReport']);
};
```

IIUC, I should be informing the test runner that the files to use are the "instrumented" ones, not the original ones, and IIUC, I should be able to do that conditionally, so that I can run `grunt test` and use my original files, or run `grunt coverage` and run the same tests on the "instrumented" ones. I do not see how to achieve this.

Even if I hard code "please just use the location of my instrumented files", I don't see much going on there. I give the command `grunt coverage`, then give `find . -mmin -1`, and nothing was generated, in particular not in `test/coverage/reports`.

**Comment by mfrasca** [comment]
is it so, that the focus is on node applications? not on browser-run javascript?

**Comment by Arkni** [comment]
Now, you need to send the coverage data from the page (html) to node. To do that, you have to use a custom bridge and inject it into the page. That bridge will send a message using `alert` (old good times :D).

Following are some steps that will help you get your coverage:

#### Folder structure

First, we choose a consistent structure:
```
.
└── coverage/
    ├── instrumented/
    └── report/
        └── ghini/
```

#### Bridge

Suppose the file name is `public/test/coverage-bridge.js`:
```js
(function(QUnit) {

  // Send messages to the parent PhantomJS process via alert! Good times!!
  function sendMessage() {
    var args = [].slice.call(arguments);
    window.alert(JSON.stringify(args));
  }

  // This will fire whenever the test suite ends
  QUnit.done(function() {

    // send coverage data if available
    if (window.__coverage__) {
      sendMessage("qunit.coverage", window.__coverage__);
    }
  });
}(QUnit));
```

#### Inject the bridge file
In your `Gruntfile` and more precisely your `qunit` task, update it to this:
```js
var path = require("path");

module.exports = function(grunt) {
  // Project configuration.
  grunt.initConfig({
    qunit: {
      files: ['test/test.html'],
      options: {
        "--web-security": "no",

        // This will inject your custom bridge, in addition to
        // the built-in bridge, into your test page
        inject: [
          path.resolve("./public/test/coverage-bridge.js"),
          require.resolve("grunt-contrib-qunit/phantomjs/bridge")
        ]
      }
    },
    // ...
  };

  // ...
};
```

#### Listen to the event:

You can then listen to that event in your `Gruntfile` as follows:
```js
grunt.event.on("qunit.coverage", function(coverage) {
  var reportPath = "coverage/report/ghini/coverage.json";

  // Create the coverage file
  grunt.file.write(reportPath, JSON.stringify(coverage));
});
```

#### Move the src file, we want to measure its coverage, around.

For that, you can use the `copy` task by installing [`grunt-contrib-copy`](https://github.com/gruntjs/grunt-contrib-copy) and configure it in your `Gruntfile`:

```js
module.exports = function(grunt) {
  // Project configuration.
  grunt.initConfig({
    // ...
    copy: {
      // Moves files around during coverage runs
      "original-to-tmp": {
        src: "public/js/cuchubo.js",
        dest: "public/js/cuchubo.tmp.js"
      },
      "instrumented-to-original": {
        src: "coverage/instrumented/js/cuchubo.js",
        dest: "public/js/cuchubo.js"
      },
      "tmp-to-original": {
        src: "public/js/cuchubo.tmp.js",
        dest: "public/js/cuchubo.js"
      }
    }
  };

  // Load plugins
  grunt.loadNpmTasks("grunt-contrib-copy");
  // ... other plugins
};
```

#### Coveralls and istanbul tasks:

```js
instrument: {
  files: "public/js/*.js",
  options: {
    lazy: true,
    basePath: "coverage/instrument/"
  }
},
storeCoverage: {
  options: {
    dir: "coverage/report"
  }
},
makeReport: {
  src: "coverage/report/**/*.json",
  options: {
    type: "lcov",
    dir: "coverage/report",
    print: "detail"
  }
},
coveralls: {
  // Options relevant to all targets 
  options: {
    force: false
  },
  qunit: {
    // LCOV coverage file
    src: "coverage/report/lcov.info",
  }
}
```

#### Configure tasks to run

```js
// Task to run tests
grunt.registerTask("test", "qunit");
grunt.registerTask("coverage", [
  "env:coverage",
  "instrument",
  "original-to-tmp",
  "instrumented-to-original",
  "test",
  "tmp-to-original",
  "storeCoverage",
  "makeReport",
  "coveralls"
]);
```

#### Finally

Your `Gruntfile` will look like this one: [`Gruntfile.js`](https://gist.github.com/Arkni/bb13c71efcba514941f0bbe00b3c45c4)

I hope these steps give you some insights on how to configure your tasks.

**Comment by Arkni** [comment]
@mfrasca I just saw your comment https://github.com/gruntjs/grunt-contrib-qunit/issues/139#issuecomment-279696762.

The answer is the two.

You still need node to build your project and run its tests in a headless browser from your terminal without the need to run a browser (`Chrome`, ...) manually and test.

**Comment by mfrasca** [comment]
this brings me a lot of steps further! thanks. I have doubts though, obviously.
- when you write »require.resolve("grunt-contrib-qunit/phantomjs/bridge")«, you really mean it literally, or is it »path.resolve("grunt-contrib-qunit/phantomjs/bridge")«, or »path.resolve("node_modules/grunt-contrib-qunit/phantomjs/bridge")«?
- where can I read about this 'inject' thing? 
- I don't understand the role of this `inject`. here, all it does is blocking for a short while and then I get the error »PhantomJS timed out, ...«. without it, PhantomJS works, but not the rest (funny '.' problem).
- where do I get 'copy'? is it part of 'grunt-contrib-qunit', or do you mean 'grunt-copy'? I grabbed the latter, but it complains it is pinned to an older version of grunt.
- the copy actions seem to add a trailing line with a single '.' (dot) to my javascript files. obviously the tests fail because of a syntax error.
- where we have <op>, the name of a copy operation, we really want 'copy:<op>', isn't it.
- when a test fails, how do I make sure the 'copy:restore-origs' is performed anyway?

**Comment by mfrasca** [comment]
p.s.: this is the [Gruntfile.js](https://gist.github.com/mfrasca/8e13309f948f492ce2401a9e3fa1b910/ba45d92474abf6033bf0aff3b69cc573f1323930), edited.
Is it really common practice?, to use 2ch indents? I always had 4.

**Comment by mfrasca** [comment]
ok, I see, 
- I had to grab `grunt-contrib-copy`, not `grunt-copy`, which adds this silly trailing line after copying the file.
- also, you really meant what you wrote: `inject: [path.resolve("test/coverage-bridge.js"), require.resolve("grunt-contrib-qunit/phantomjs/bridge")],`
- I now have a json file in the coverage reports, so I guess I should now be almost there.

I still wonder how to make sure a failing test will not break my files... or do I first run the tests before running them with coverage. so if they fail it will stop on the first run, before starting shuffling them around.

**Comment by mfrasca** [comment]
really almost there. 
`Running "coveralls:qunit" (coveralls) task`
- we did not need the `storeCoverage` task, because this is already done by the `qunit.coverage` callback,
- coveralls needs the report generated by `makeReport`,

the problems now are possibly not any more tied to grunt or qunit. and I am looking forward to contributing some documentation for your program. ... :100:

**Comment by mfrasca** [comment]
ok, works, now down to describing it for others (and for forgetful myself).
you'll see a pull request soon, I hope, and I hope you will find it useful.

**Comment by Arkni** [comment]
Sorry for not responding soon, I was a little busy.

I see, you're almost there :+1: 

I'll try to answer some of your questions.

> Is it really common practice?, to use 2ch indents? I always had 4.

Nope, I just used to that style.

> when you write »require.resolve("grunt-contrib-qunit/phantomjs/bridge")«, you really mean it literally, or is it »path.resolve("grunt-contrib-qunit/phantomjs/bridge")«, or »path.resolve("node_modules/grunt-contrib-qunit/phantomjs/bridge")«?

`require.resolve("grunt-contrib-qunit/phantomjs/bridge")` and `path.resolve("node_modules/grunt-contrib-qunit/phantomjs/bridge.js")` (notice the `.js` extension) are the same in your case where `node_modules` and `Gruntfile` are in the same directory. But in case you don't have a clue in where the modules `grunt-contrib-qunit` is installed, you have to use `require.resolve("grunt-contrib-qunit/phantomjs/bridge")` which will return the full filepath of the bridge file.

> I had to grab grunt-contrib-copy, not grunt-copy, which adds this silly trailing line after copying the file. (you had written the correct name, I missed it).

Yep, that's it.

> also, you really meant what you wrote: inject: [path.resolve("test/coverage-bridge.js"), require.resolve("grunt-contrib-qunit/phantomjs/bridge")], (but I don't see much relationship with its documentation)

That documentation should be updated to be the same as [`grunt-lib-phantomjs#inject`](https://www.npmjs.com/package/grunt-lib-phantomjs#inject). I forgot to update it when updating the linked one.

> we did not need the storeCoverage task, because this is already done by the qunit.coverage callback,

Ahh, yes. You're right!

> I still wonder how to make sure a failing test will not break my files... or do I first run the tests before running them with coverage. so if they fail it will stop on the first run, before starting shuffling them around.

I'll take a look.
